internal/daemon/worker/proxy/proxydialer.go

```diff
@@ -19,19 +22,70 @@ var GetEndpointDialer = directDialer
 
 // directDialer returns a ProxyDialer which tcp dials directly to the provided
 // endpoint.
-func directDialer(ctx context.Context, endpoint string, _ string, _ proto.Message, _ interface{}) (*ProxyDialer, error) {
+func directDialer(ctx context.Context, endpoint string, _ string, _ proto.Message, _ interface{}, opt ...Option) (*ProxyDialer, error) {
 	const op = "proxy.directDialer"
 	if len(endpoint) == 0 {
 		return nil, errors.New(ctx, errors.InvalidParameter, op, "endpoint is empty")
 	}
-	d, err := NewProxyDialer(ctx, func(opt ...Option) (net.Conn, error) {
+	opts := GetOpts(opt...)
+	dnsServer := opts.WithDnsServer
+	parsedDnsServer, err := url.Parse(dnsServer)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	d, err := NewProxyDialer(ctx, func(dialerOpt ...Option) (net.Conn, error) {
+		if dnsServer != "" {
+			host, port, err := net.SplitHostPort(endpoint)
+			if err != nil {
+				return nil, errors.Wrap(ctx, err, op, errors.WithMsg("error splitting host/port"))
+			}
+			_, err = netip.ParseAddr(host)
+			if err == nil || host == "localhost" {
+				// It's an IP address or something we definitely can't resolve,
+				// continue
+				goto dial
+			}
```

Comment:
Could we invert this if statement and remove the goto?
```suggestion
			// It's an IP address or something we definitely can't resolve,
			// continue
			if err != nil && host != "localhost" {
				dnsClient := new(dns.Client)
				...
			}
```

Reply:
Sure, I can